typeclass-experiments/src/Gen/Rust.hs

```diff
@@ -0,0 +1,161 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Gen.Rust where
```

Comment:
Yeah, I think mapping directly to Text/Doc is not a good idea, we need to make this very clean, especially when we start mapping to actual Rust builtins, which means we need to manage module/package information and all that stuff.

Concatenating string at this stage seems like a foot gun. But it doesn't really change much, we just parametrize the generator with a Rust DSL type.

Reply:
Well, this is kind of an attempt to get started on an eDSL that does that I suppose.

I think `Doc` is fine in the sense that: When we look at a schema and try to generate some code for it, we don't need anything outside of that schema to generate the textual representation of the code, so we *should* be able to generate the prettily-formatted `Doc` straightaway. 

But since, in the course of generating code for a schema, we might also need to generate an import statement (or whatever), Doc can't be the _only_ thing that the generator outputs. We need generator functions to return something that's (morally) `(Doc,[Import])`. We probably can't generate `Doc` representations of imports directly because we'll want to do things like combine imports from the same module into one import statement & eliminate redundant imports & so on. 

We will need a set of utility functions for each target language for formatting though. Like, we'll need a DSL (well an eDSL) for bracketing and indentation conventions particular to rust. 

The advantage of this approach going forward, btw, is that we can write little modular code generators vs doing everything in massive case expressions on the `Source` like PS-bridge does. We *could* just do that, but it would be really unreadable and hard to debug and maintain - I think that's one lesson we learned from ps-bridge. 

I think we _need_ something like this approach; once we start implementing dozens (hundreds?) of opaque types (e.g. for the ledger), it will become a nightmare to handle every special case in a monolithic function. 

Also, we should be able to specify a finite number of patterns such that if we have a corresponding generator for each then we know we have a complete generator for a target language.

Reply:
Not sure we're talking about the same thing. Consider a Rust DSL:

```
module LambdaBuffers.Rust where

data Module = Module {
  tyDefs :: TyDef,
  traitDefs :: TraitDef
}

data TyDef = TyDef {
  tyName :: String,
  ...
  tyBody :: tyBody
}

data Ty = ... | ... | TyRef TyRef

data Val = ValAbs .. | ValApp Val Val | ValRef ...
```

I would like to formulate what Rust IS from our point of view, to a very coarse level of detail, such that we map from LB to that ADT first. This is very informative, separates concerns such that it makes explicit the following to tasks:
1. How are LB constructs mapped to Rust constructs,
2. How are Rust constructs serialized to Text.

I feel this to be crucial to sustainability, opens up a neat testing opportunity, makes clear the different tasks the codegen does.

Bare in mind that Pat is still used to construct a `Rust.Module` from `LB.Module`, but you're not concatenating Text and messing with formatting at this point. And also note that you can perform 'fresh' variable and function naming at this stage, such that when you get to transforming `Rust.Module` to `Text` you don't have to do any heavy lifting.

Reply:
>The advantage of this approach going forward, btw, is that we can write little modular code generators vs doing everything in massive case expressions on the Source like PS-bridge does. We could just do that, but it would be really unreadable and hard to debug and maintain - I think that's one lesson we learned from ps-bridge.

I'm completely sold on that.